packages/stream-text/src/index.ts

```diff
@@ -32,96 +50,279 @@ export interface StreamTextOptions extends ChatOptions {
+        const choice = chunk.choices[0]
+
+        // mark this time as non-text output if is has toolcalls
+        if (choice.delta.tool_calls) {
+          shouldOutputText = false
+        }
+
+        const { delta, finish_reason, index, ...rest } = choice
+        const choiceSnapshot: StreamTextChoice = step.choices[index] ??= {
+          finish_reason,
+          index,
+          message: {
+            role: 'assistant',
+          },
+        }
+
+        if (finish_reason !== undefined) {
+          choiceSnapshot.finish_reason = finish_reason
+
+          if (finish_reason === 'length') {
+            throw new Error('length exceeded')
+          }
+
+          if (finish_reason === 'content_filter') {
+            throw new Error('content filter')
+          }
+        }
+
+        Object.assign(choiceSnapshot, rest)
+
+        const { content, refusal, tool_calls, ...rests } = delta
+        const message = choiceSnapshot.message
+        Object.assign(message, rests)
+
+        if (refusal !== undefined) {
+          // eslint-disable-next-line ts/strict-boolean-expressions
+          message.refusal = (message.refusal || '') + refusal
+        }
+
+        if (content !== undefined) {
+          // eslint-disable-next-line ts/strict-boolean-expressions
+          message.content = (message.content || '') + content
+          shouldOutputText && textCtrl?.enqueue(content)
+        }
+
+        for (const { function: fn, id, type } of tool_calls || []) {
+          message.tool_calls ??= {}
+
+          const toolCall = message.tool_calls[id] ??= {
+            function: {
+              arguments: '',
+              name: fn.name,
+            },
+          } as StreamTextToolCall
+          toolCall.id = id
+          toolCall.type = type
+          toolCall.function.arguments += fn.arguments
+        }
+
+        /**
+         * check if toolcall is ended
+         */
+        const state = choiceState[index] ??= {
+          calledToolCallIDs: new Set(),
+          currentToolID: null,
+          endedToolCallIDs: new Set(),
+          index,
+          toolCallErrors: {},
+          toolCallResults: {},
+        }
+        // eslint-disable-next-line ts/strict-boolean-expressions
+        if (finish_reason) {
+          // end choice will end the current toolcall too
+          if (state.currentToolID !== null) {
+            endToolCall(state, state.currentToolID)
+          }
+        }
+        for (const toolCall of delta.tool_calls || []) {
+          // new toolcall
+          if (state.currentToolID !== null && state.currentToolID !== toolCall.id) {
+            endToolCall(state, state.currentToolID)
+          }
+
+          state.calledToolCallIDs.add(toolCall.id)
+          state.currentToolID = toolCall.id
+        }
+      },
+    })),
+    )
+
+    shouldOutputText && step.messages.push({
+      content: step.choices[0].message.content ?? '',
+      refusal: step.choices[0].message.refusal,
```

Comment:
Is there any chance this `[0]` index will fail?

Reply:
Not likely, but we can add a guard